File: src/main/Python/nowcoder/S38_TreeDepth.py
import PrintTreeLayer
class TreeDepth:
    def TreeDepth(self, root):
        ld = self.TreeDepth(root.left)
        rd = self.TreeDepth(root.right)
        if ld > rd:
            return ld +1
        else: return rd +1

if __name__ == '__main__':
    test = TreeDepth()
    p = PrintTreeLayer.PrintTreeLayer()
    array = [1, 2, 3, 4, 5, 6, 7, 8]
    root = p.arrayToTree(array, 0)
    print(test.TreeDepth(root))

File: src/main/java/nowcoder/S38_TreeDepth.java
package nowcoder;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author LiuZhiguo
 * @date 2020/2/4 14:28
 */
public class S38_TreeDepth {
    //递归方法
    /*public int TreeDepth(TreeNode root){
        if (root == null)
            return 0;
        int ld = TreeDepth(root.left);
        int rd = TreeDepth(root.right);
        return ld > rd ? ld+1 : rd+1;
    }*/
    //非递归方法，层次遍历
    public int TreeDepth(TreeNode root){
        if (root == null)
            return 0;
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        int depth = 0;
        int count = 0;
        int layerCount = 1;
        queue.offer(root);
        TreeNode out;
        while (!queue.isEmpty()){
            out = queue.poll();
            count++;
            if (out.left != null)
                queue.offer(out.left);
            if (out.right != null)
                queue.offer(out.right);
            if (count == layerCount){
                depth++;
                layerCount = queue.size();
                count = 0;
            }
        }
        return depth;
    }
    public static void main(String[] args){
        S38_TreeDepth s38 = new S38_TreeDepth();
        PrintTreeLayer pt = new PrintTreeLayer();
        Integer[] array = {1, 2, 3, 4, 5, 6, 7, 8};
        TreeNode root = pt.arrayToTree(array, 0);
        System.out.println(s38.TreeDepth(root));
    }
}
